### PR TITLE
Small issues fixed.

### LINKED ISSUE #6158: Pointers for writing ad-hoc operators involving 4D tensors (conv-like) on the GPU
So, I need to implement the following operation. Assume that we have a standard 2D convolution of a batch of images `I` of size `(N, C, W, H)` with a kernel weights `W` of size `(O, C, D, D)` (will assume kernels are square). Now let's write the convolution in the following form:
```
# This will be `(N*T) x (C*D1*D2)`
i2 = image2neigbs(I)
# This will be `(C*D1*D2)xO`
w2 = W.dimshuffle(1, 2, 3, 0).reshape((-1, O))
# This will be `(N*T)xO`
out = dot(i2, w2)
```
So far so good, however I need to calculate
```
# This will be (C*D1*D2)x(C*D1*D2)
X = dot(i2.T, i2)
```
I've tested the naive way to do it as wit the `image2neigbs` and here is the profiling:
```
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Class name>
  83.3%    83.3%      39.646s       6.61e-01s     C       60      12   theano.tensor.nnet.neighbours.Images2Neibs
   6.4%    89.8%       3.068s       9.59e-03s     C      320      64   theano.gpuarray.basic_ops.GpuFromHost
   4.2%    94.0%       1.999s       1.11e-02s     C      180      36   theano.gpuarray.blas.GpuDot22
   1.4%    95.4%       0.660s       2.75e-03s     C      240      48   theano.gpuarray.basic_ops.HostFromGpu
   0.9%    96.3%       0.428s       3.89e-03s     C      110      22   theano.gpuarray.dnn.GpuDnnConvGradI
   0.8%    97.1%       0.377s       5.89e-04s     C      640     128   theano.gpuarray.basic_ops.GpuReshape
   0.6%    97.7%       0.298s       1.19e-02s     C       25       5   theano.gpuarray.blas.GpuGemm
   0.4%    98.1%       0.188s       1.71e-03s     C      110      22   theano.gpuarray.basic_ops.GpuJoin
   0.4%    98.5%       0.176s       9.78e-04s     C      180      36   theano.gpuarray.dnn.GpuDnnConv
   0.4%    98.8%       0.172s       1.11e-04s     C     1550     310   theano.gpuarray.elemwise.GpuElemwise
   0.3%    99.2%       0.157s       6.03e-04s     C      260      52   theano.gpuarray.elemwise.GpuCAReduceCuda
   0.2%    99.4%       0.113s       8.38e-04s     Py     135      27   theano.gpuarray.linalg.GpuCusolverSolve
   0.2%    99.6%       0.103s       2.89e-04s     C      355      71   theano.gpuarray.dnn.GpuDnnReduction
   0.2%    99.8%       0.102s       1.69e-03s     C       60      12   theano.gpuarray.dnn.GpuDnnConvGradW
   0.0%    99.9%       0.021s       8.65e-05s     C      245      49   theano.gpuarray.basic_ops.GpuAlloc
   0.0%    99.9%       0.020s       5.51e-05s     C      365      73   theano.gpuarray.basic_ops.GpuAllocEmpty
   0.0%    99.9%       0.012s       6.07e-04s     C       20       4   theano.gpuarray.dnn.GpuDnnPoolGrad
   0.0%    99.9%       0.005s       5.53e-06s     C      940     188   theano.gpuarray.basic_ops.GpuContiguous
   0.0%   100.0%       0.005s       2.12e-05s     C      230      46   theano.gpuarray.subtensor.GpuIncSubtensor
   0.0%   100.0%       0.004s       3.84e-04s     C       10       2   theano.gpuarray.pool.GpuDownsampleFactorMaxGradGrad
   ... (remaining 18 Classes account for   0.04%(0.02s) of the runtime)
```
As can be seen the `Images2Neibs` takes 83% of the compute time. At this stage I want to implement the whole operation as a proper GPU operator. However, I was wondering if anyone could point me out how to actually optimize this (e.g. if there are any code for convolutions which are not cuDNN) and how to manuver the 4D tensor in cuda (how to pass the sizes of the array to the funciton etc...).
I know this is noones job to do but any help would be most appriciated.

### PR DESCRIPTION
**Pull request by botev**
Added convenience function for `solve_symmetric`. 
Added correct handling of `lower_triangular` and `upper_triangular` in the GpuSolve Operator.
Moved the Op lifter for Images2Neibs in `opt.py` such that it is included in the optimisation passes. Closes #6158

### CODE REVIEW

**Comment by abergeron** [comment]
I'm ok for the change in neighbours, but I have no idea about the linalg change.

**Comment by botev** [comment]
@abergeron Just to clear it up. For the symmetric case the first call `cusolver.cusolver.DnSpotrf` calculates the Cholesky decomposition, by writing in the Upper or Lower part of `A_ptr` depending on the flag that you pass. Then the `cusolverDnSpotrs` performs the actual solving in of the linear system using a Lower or Upper triangular input. When we have `upper` or `lower` triangular structure basically the first step is not needed and one can just directly invoke `cusolverDnSpotrs`.  That's it.

**Comment by botev** [comment]
@abergeron @lamblin I reversed the `cusolver` modification. Sorry for the inconvenience, that probably should be a separate PR. I've posted to `skcuda` team to see if there exists a procedure for solving a linear system directly from the Cholesky.
Just to clear up the error is that `cusolverDnSpotrs` given the cholesky `L` solves:
`(L L^T)x = b`
while we need a function which solves
`Lx=b`
So again my mistake sorry for wasting time. The Image2Neigbs should be fine to merge.

**Comment by notoraptor** [comment]
The PR must also be rebased.

**Comment by botev** [comment]
@abergeron Done and @notoraptor Done. I just kept the added convenience function for `solve_symmetric` I think that is reasonable.

**Comment by notoraptor** [comment]
@botev Yes, you can keep `solve_symmetric`.

Can you also rebase the PR ? There is a conflict because another PR was recently merged.

**Comment by botev** [comment]
@notoraptor Any pointers on how to rebase it? I removed the conflict with merging I think, although I told git to rebase.

**Comment by notoraptor** [comment]
@botev This is some general documentations about rebasing:
- https://help.github.com/articles/about-git-rebase/
- https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

**Comment by notoraptor** [comment]
You have flake8 errors.

**Comment by botev** [comment]
@notoraptor Yeah I always get that from the first few times... You guys should get something like `rustfmt` or so... this whole flake8 and pep8 are so annoying otherwise.

File: theano/gpuarray/neighbours.py [hunk 1]
```diff
@@ -577,12 +576,4 @@ def c_code(self, node, name, inp, out, sub):
...
-        Op.perform(self, node, inp, out, params)
-
-
-@register_opt('fast_compile')
-@op_lifter([Images2Neibs])
-@register_opt2([Images2Neibs], 'fast_compile')
-def local_gpua_images2neibs(op, context_name, inputs, outputs):
-    if op.mode in ['valid', 'half', 'full', 'ignore_borders', 'wrap_centered']:
-        return GpuImages2Neibs(op.mode)
+        Op.perform(self, node, inp, out, params)
```
Comment (notoraptor) [inline theano/gpuarray/neighbours.py]: Also, add a newline after this line.

**Comment by juancamilog** [comment]
triangular solves can be done on cublas (through skcuda.cublas.cublasStrsm) as done in #4007

**Comment by botev** [comment]
@juancamilog Yes actually that seems about right. I will still leave it for another PR, but that might indeed be what is required for the solver.

**Comment by botev** [comment]
@notoraptor This should be good to merge now.

**Pull request merged**